docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+of the custom metric API, but also of the external metric API, preventing the deployment of alternative handlers for
+these API extension points. 
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole)
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole)
+4. New feature flag promoted and eventually removed
+5. (Outside the scope of this proposal) use of HVPA removed for other workloads and the preexisting HVPA flag removed
+
+## Discussion and Limitations
+#### Prometheus-metrics-adapter precludes the use of sources of custom metrics
+With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension
+point is occupied by prometheus-metrics-adapter.
```

Comment:
Why do you mention this explicitly? Is this (only one API extension can serve custom metrics a problem or will it become one?

Reply:
It imposes constraint on future design choices. Currently, we have the freedom to register any custom metrics provider. After the change, if a need arises for custom metrics which cannot be provided by the prometheus-metrics-adaper, we'll be in a position where we need to figure out a way to provide them, without disrupting the metrics required by the proposed HPA.

Reply:
Yeah, sure, but what is the alternative now? I don't think there is any since the approach relies on the custom metrics. Hence, we have to go down this path anyways, I assume?

Reply:
The main point is to keep this (occupying the custom metrics extension point) in mind as part of the cost of the proposed approach.

There are more specific concerns, but from the perspective of this review, I think they are not important enough to consider. For instance, an alternative implementation of the proposed approach is to have custom metrics provided by a small custom component, instead of prometheus-metrics-adapter. That alternative approach has the benefit that it could easily coexist with another custom metrics provider by selectively forwarding queries to it. But I think that's of negligible importance at present. It suffices to note that the proposed approach hinders future use of other sources of custom metrics.

Reply:
Yes. Just for the record, maybe you can put this

> hat alternative approach has the benefit that it could easily coexist with another custom metrics provider by selectively forwarding queries to it.

into the document then. I doubt that we will ever need it, but just in case (and since we now discussed it here). Thanks.

Reply:
Done. Will be available in next push.